packages/firestore/src/core/transaction.ts

```diff
@@ -64,10 +64,11 @@ export class Transaction {
     this.ensureCommitNotCalled();
 
     if (this.mutations.length > 0) {
-      throw new FirestoreError(
+      this.lastTransactionError = new FirestoreError(
         Code.INVALID_ARGUMENT,
         'Firestore transactions require all reads to be executed before all writes.'
       );
+      throw this.lastTransactionError;
```

Comment:
It is possible that the same error gets thrown twice: once is here when we detect it, and throw again when we call commit() since it is stored into this.lastTransactionError. Maybe just throw it inside commit().

Reply:
It's a good thought, but the [TransactionRunner will handle this error using only one of the two possible code paths](https://github.com/firebase/firebase-js-sdk/blob/e2a90bf678eb6fa505f8b2f627e03cff622607b5/packages/firestore/src/core/transaction_runner.ts#L62-L77):
If the user's transaction function awaits the failed get/lookup, then the userPromise will reject and the transaction runner will handle it and commit will not be run. Otherwise, if the user's transaction function does not await the failed get/lookup, then the userPromise will not reject. The commit will be run but it will fail with this error.

I want to continue throwing in this scenario in case any customers are relying on that behavior. Not likely, but it is an API change to stop throwing in this case.